File: in_Java/1417 Reformat The String.java
class Solution {
    public String reformat(String s) {
        int num_c = 0;
        int num_n = 0;
        for(char c : s.toCharArray()){
            if(c >= 'a' && c <= 'z') num_c++;
            else num_n++;
        }
        if(Math.abs(num_c - num_n) > 1) return "";
        char[] res = new char[num_c + num_n];
        char[] allchar = new char[num_c];
        char[] allnum = new char[num_n];
        int i = 0, j = 0;
        for(char c : s.toCharArray()){
            if(c >= 'a' && c <= 'z') allchar[i++] = c;
            else allnum[j++] = c;
        }
        int c_id = 0, n_id = 0, id = 0;
        if(num_c > num_n){
            res[id++] = allchar[c_id++];
            while(id < res.length){
                res[id++] = allnum[n_id++];
                res[id++] = allchar[c_id++];
            }
        }else if(num_c < num_n){
            res[id++] = allnum[n_id++];
            while(id < res.length){
                res[id++] = allchar[c_id++];
                res[id++] = allnum[n_id++];
            }
        }else{
            while(id < res.length){
                res[id++] = allchar[c_id++];
                res[id++] = allnum[n_id++];
            }
        }
        return String.valueOf(res);
    }
}

File: in_Python/1417 Reformat The String.py
class Solution(object):
    def reformat(self, s):
        """
        :type s: str
        :rtype: str
        """
        dc = []
        lc = []
        for si in s:
            if si.isdigit():
                dc.append(si)
            else:
                lc.append(si)
        if abs(len(dc) - len(lc)) > 1:
            return ""
        res = ''
        if len(dc) > len(lc):
            for i in range(len(lc)):
                res += dc[i]
                res += lc[i]
            res += dc[-1]
        elif len(dc) < len(lc):
            for i in range(len(dc)):
                res += lc[i]
                res += dc[i]
            res += lc[-1]
        else:
            for i in range(len(dc)):
                res += lc[i]
                res += dc[i]
        return res